src/providers/message-provider.tsx

```diff
@@ -0,0 +1,472 @@
+  const [keys, setKeys] = useState<Keys>();
+  const [messageService, setMessageService] = useState<any>();
+  const [directMessageBuffer, setDirectMessageBuffer] = useState<DirectMessage[]>([]);
+  const [publicMessageBuffer, setPublicMessageBuffer] = useState<PublicMessage[]>([]);
+  const [isCommunityCreated, setIsCommunityCreated] = useState(false);
+  const [replacedPublicMessagesBuffer, setReplacedPublicMessagesBuffer] = useState<string[]>([]);
+  const [isDirectChatCreated, setIsDirectChatCreated] = useState(false);
+  const [replacedDirectMessagesBuffer, setReplacedDirectMessagesBuffer] = useState<string[]>([]);
+
+  useEffect(() => {
+    if (chat.channels.length !== 0) {
+      setIsCommunityCreated(true);
+    }
+    if (chat.directMessages.length !== 0) {
+      setIsDirectChatCreated(true);
+    }
+  }, [chat.channels, chat.directMessages]);
+
+  useEffect(() => {
+    window.addEventListener("createMSInstance", createMSInstance);
+
+    return () => {
+      window.removeEventListener("createMSInstance", createMSInstance);
+    };
+  }, []);
+
+  useEffect(() => {
+    if (!window.messageService && keys?.priv) {
+      const messageServiceInstance = initMessageService(keys);
+      setMessageService(messageServiceInstance);
+    }
+  }, [keys]);
+
+  useEffect(() => {
+    if (activeUser) {
+      getNostrKeys(activeUser);
+    }
+  }, [activeUser]);
+
+  const createMSInstance = (e: Event) => {
+    const detail = (e as CustomEvent).detail as NostrKeysType;
+    const messageServiceInstance = initMessageService(detail);
+    setMessageService(messageServiceInstance);
+  };
+
+  const getNostrKeys = async (activeUser: ActiveUser) => {
+    const profile = await getProfileMetaData(activeUser.username);
+    const noStrPrivKey = getPrivateKey(activeUser.username);
+    const keys = {
+      pub: profile.nsKey,
+      priv: noStrPrivKey
+    };
+    setKeys(keys);
+  };
+
+  //Listen for events in an interval.
+  useEffect(() => {
+    if (!messageServiceReady) return;
+
+    const timer = setTimeout(
+      () => {
+        messageService?.listen(
+          chat.channels.map((x) => x.id),
+          Math.floor((since || Date.now()) / 1000)
+        );
+        setSince(Date.now());
+      },
+      since === 0 ? 500 : 10000
+    );
+
+    return () => {
+      clearTimeout(timer);
+    };
+  }, [since, messageServiceReady, messageService, chat.channels]);
+
+  // // Ready state handler
+  const handleReadyState = () => {
+    setMessageServiceReady(true);
+  };
+
+  useEffect(() => {
+    messageService?.removeListener(MessageEvents.Ready, handleReadyState);
+    messageService?.addListener(MessageEvents.Ready, handleReadyState);
+
+    return () => {
+      messageService?.removeListener(MessageEvents.Ready, handleReadyState);
+    };
+  }, [messageServiceReady, messageService]);
+
+  // Profile update handler
+  const handleProfileUpdate = (data: Profile[]) => {
+    props.addProfile(data);
+  };
```

Comment:
The addProfile is a reducer function from store. I want to call this in an event listener. Therefore, I created a wrapper function for it to provide in event listener so that the code remains same for all event listeners.